src/Simplex/Chat.hs

```diff
@@ -6012,15 +6072,82 @@ createGroupFeatureChangedItems user cd ciContent GroupInfo {fullGroupPreferences
 sameGroupProfileInfo :: GroupProfile -> GroupProfile -> Bool
 sameGroupProfileInfo p p' = p {groupPreferences = Nothing} == p' {groupPreferences = Nothing}
 
-createInternalChatItem :: forall c d m. (ChatTypeI c, MsgDirectionI d, ChatMonad m) => User -> ChatDirection c d -> CIContent d -> Maybe UTCTime -> m ()
-createInternalChatItem user cd content itemTs_ = do
+createInternalChatItem :: (ChatTypeI c, MsgDirectionI d, ChatMonad m) => User -> ChatDirection c d -> CIContent d -> Maybe UTCTime -> m ()
+createInternalChatItem user cd content itemTs_ =
+  createInternalItemsForChatsB user itemTs_ (Identity (cd, content)) >>= \case
+    [Right aci] -> toView $ CRNewChatItem user aci
+    [Left e] -> throwError e
+    rs -> throwChatError $ CEInternalError $ "createInternalChatItem: expected 1 result, got " <> show (length rs)
+
+createInternalChatItems :: (ChatTypeI c, MsgDirectionI d, ChatMonad' m) => User -> Maybe UTCTime -> ChatDirection c d -> [CIContent d] -> m [Either ChatError AChatItem]
+createInternalChatItems user itemTs_ cd ciContents = do
+  let dirsCIContents = map (cd,) ciContents
+  createInternalItemsForChatsB user itemTs_ dirsCIContents
+
+-- remove
+createInternalItemsForChats ::
+  (ChatTypeI c, MsgDirectionI d, ChatMonad' m) =>
+  User ->
+  Maybe UTCTime ->
+  [(ChatDirection c d, CIContent d)] ->
+  m [Either ChatError AChatItem]
+createInternalItemsForChats = createInternalItemsForChatsB
+
+-- TODO refactor to return t (Either ChatError AChatItem)
+createInternalItemsForChatsB ::
+  forall c d m t.
+  (ChatTypeI c, MsgDirectionI d, ChatMonad' m, Traversable t) =>
+  User ->
+  Maybe UTCTime ->
+  t (ChatDirection c d, CIContent d) ->
+  m [Either ChatError AChatItem]
```

Comment:
```suggestion
  t (Either ChatError (ChatDirection c d, CIContent d)) ->
  m (t (Either ChatError AChatItem))
```

Reply:
the point of this function to traverse over any structure, skipping over previous errors - then you can compose these operations.

Reply:
... and to apply composition to lists and to single items inside Identity

Reply:
or probably it should be:

t (Either ChatError (ChatDirection c d, [CIContent d])) ->
  m (t (Either ChatError [AChatItem]))
  
assuming you don't fail if creating one item fails and only fail on operation that applies to ChatDirection c d

Reply:
so the list size can change, but "t" still won't change